python/veles/schema/model.py

```diff
@@ -12,47 +12,66 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from copy import copy
+
+import six
+
 from veles.compatibility import pep487
+from veles.proto.exceptions import SchemaError
+from . import fields
 
 
 class Model(pep487.NewObject):
+    fields = []
+
     def __init__(self, **kwargs):
-        if any(i not in [field.name for field in self.fields] for i in kwargs):
-            raise TypeError('got an unexpected keyword argument')
-        for field_name, field in kwargs.items():
-            setattr(self, field_name, field)
+        for field in self.fields:
+            if field.name in kwargs:
+                setattr(self, field.name, kwargs.pop(field.name))
+            else:
+                setattr(self, field.name, copy(field.default))
+        if kwargs:
+            raise TypeError('got an unexpected keyword argument {}'.format(
+                kwargs.popitem()[0]))
 
     def __init_subclass__(cls, **kwargs):
         super(Model, cls).__init_subclass__(**kwargs)
 
-        cls.fields = []
-        for klass in cls.__bases__ + (cls,):
-            for attr_name, attr in klass.__dict__.items():
-                try:
-                    attr.add_to_class(cls)
-                except AttributeError:
-                    pass
+        cls.fields = super(cls, cls).fields[:]
+        for attr in cls.__dict__.values():
+            if isinstance(attr, fields.Field):
+                cls.fields.append(attr)
 
-    def to_dict(self):
+    def dump(self):
         val = {}
         for field in self.fields:
-            val[field.name] = field.__get__(self)
+            val[six.text_type(field.name)] = field.dump(field.__get__(self))
         return val
 
     @classmethod
-    def from_dict(cls, val):
-        """Override if your class overrides __init__"""
-        return cls(**val)
-
-    @classmethod
     def load(cls, val):
-        return cls.from_dict(val)
-
-    def dump(self, packer):
-        return packer.pack(self.to_dict())
+        if not isinstance(val, dict):
+            raise SchemaError('a serialized model must be a dict')
+        val = dict(val)
+        res = cls.__new__(cls)
+        for field in cls.fields:
```

Comment:
The idea, like for pickle, is to have this work even if __init__ takes different arguments.  But I'm not really convinced either way...